File: Tutorial/Tutorial/target.cpp
#include "target.h"

int serialize(const target &data, char *buf) {
	uint16_t *ids = (uint16_t*)buf;

	*ids = data.cam_id;    ids++;
	*ids = data.target_id; ids++;

	double *tra = (double*)ids;

	*tra = data.pos_x; tra++;
	*tra = data.pos_y; tra++;
	*tra = data.pos_z; tra++;

	*tra = data.rot_x; tra++;
	*tra = data.rot_y; tra++;
	*tra = data.rot_z; tra++;
	*tra = data.rot_w; tra++;

	return 0;
}

int deserialize(char *buf, target &data) {
	uint16_t *ids = (uint16_t*)buf;

	data.cam_id = *ids;    ids++;
	data.target_id = *ids; ids++;

	double *tra = (double*)ids;

	data.pos_x = *tra; tra++;
	data.pos_y = *tra; tra++;
	data.pos_z = *tra; tra++;

	data.rot_x = *tra; tra++;
	data.rot_y = *tra; tra++;
	data.rot_z = *tra; tra++;
	data.rot_w = *tra; tra++;

	return 0;
}

int sendData(const target &data) {
	/* Build packet */

	char buf[target::SIZE];
	serialize(data, buf);

	/* Unicast to devices */

	return 1;
}

target::target(uint16_t c, uint16_t t, double x, double y, double z, double rx, double ry, double rz, double rw)
	:cam_id(c), target_id(t), pos_x(x), pos_y(y), pos_z(z), rot_x(rx), rot_y(ry), rot_z(rz), rot_w(rw)
{
	
}

File: Client/Assets/Target.cs
using System;

public struct Target {

  public const int SIZE
    = 2 * sizeof(ushort)
    + 7 * sizeof(double);

  public ushort cam_id;
  public ushort target_id;

  public double pos_x;
  public double pos_y;
  public double pos_z;

  public double rot_x;
  public double rot_y;
  public double rot_z;
  public double rot_w;

  public void Deserialize(byte[] buffer) {
    int index = 0;

    this.cam_id = BitConverter.ToUInt16(buffer, index);
    index += sizeof(ushort);

    this.cam_id = BitConverter.ToUInt16(buffer, index);
    index += sizeof(ushort);

    /* Position */

    this.pos_x = BitConverter.ToDouble(buffer, index);
    index += sizeof(double);

    this.pos_y = BitConverter.ToDouble(buffer, index);
    index += sizeof(double);

    this.pos_z = BitConverter.ToDouble(buffer, index);
    index += sizeof(double);

    /* Rotation */

    this.rot_x = BitConverter.ToDouble(buffer, index);
    index += sizeof(double);

    this.rot_y = BitConverter.ToDouble(buffer, index);
    index += sizeof(double);

    this.rot_z = BitConverter.ToDouble(buffer, index);
  }
}
